lib/shelley/src/Cardano/Wallet/Shelley/Network.hs

```diff
@@ -0,0 +1,508 @@
+import Network.Socket
+    ( AddrInfo (..), Family (..), SockAddr (..), SocketType (..) )
+
+import qualified Codec.Serialise as CBOR
+import qualified Network.Socket as Socket
+import qualified Ouroboros.Network.Block as Abstract
+
+--------------------------------------------------------------------------------
+--
+-- Stateless Interface
+--
+-- This is a first attempt at using the ChainSync mini protocol in a stateless
+-- manner. Note that this is a bit unfortunate as the
+
+-- | A stateless interface for dealing with the Haskell 'NetworkClient'
+--
+-- See 'newNetworkClient' and 'connectClient' methods to construct a
+-- 'NetworkLayer':
+--
+--     let nodeId = CoreNodeId 0
+--     let addr = localSocketAddrInfo (localSocketFilePath nodeId)
+--     let params = ChainParameters
+--          { epochSlots = EpochSlots 21600
+--          , protocolMagic = ProtocolMagicId 1097911063
+--          }
+--     (client, network) <- newNetworkClient nullTracer params
+--
+--     -- Launch the stateful connection in a separate thread
+--     void $ forkIO $ connectClient client dummyNodeToClientVersion addr
+--
+--     -- Use the network layer to send requests to the node
+--     response <- runExceptT $ nextBlocks network genesisPoint
+--     point <- networkTip network
+--     response' <- runExceptT $ nextBlocks network point
+--
+data NetworkLayer m = NetworkLayer
+    { nextBlocks :: Point -> ExceptT ErrNextBlocks m [Block]
+    , networkTip :: m Point
+    }
+
+-- | This translates the 'NetworkLayer' interface from the wallet backend.
+data ChainSyncRequest m
+    = ReqNextBlocks Point (Either ErrNextBlocks [Block] -> m ())
+    | ReqNetworkTip (Point -> m ())
+
+-- | What is considered errors in the 'NetworkLayer'
+data ErrNextBlocks
+    = ErrNextBlocksNoIntersection
+    | ErrNextBlocksRollBack
+    deriving (Show)
+
+
+--------------------------------------------------------------------------------
+--
+-- Concrete Types
+
+type Block = ByronBlockOrEBB ByronConfig
+
+type Point = Abstract.Point Block
+
+type Tx = GenTx Block
+
+data ChainParameters = ChainParameters
+    { epochSlots :: EpochSlots
+        -- ^ Number of slots per epoch.
+    , protocolMagic :: ProtocolMagicId
+        -- ^ Protocol magic (e.g. mainnet=764824073, testnet=1097911063)
+    }
+
+-- | Type representing a network client running two mini-protocols to sync
+-- from the chain and, submit transactions.
+type NetworkClient m = OuroborosApplication
+    'InitiatorApp
+        -- ^ Initiator ~ Client (as opposed to Responder / Server)
+    (SockAddr, SockAddr)
+        -- ^ An identifier for the peer: here, a local and remote socket.
+    NodeToClientProtocols
+        -- ^ Specifies which mini-protocols our client is talking.
+        -- 'NodeToClientProtocols' allows for two mini-protocols:
+        --  - Chain Sync
+        --  - Tx submission
+    m
+        -- ^ Underlying monad we run in
+    ByteString
+        -- ^ Concrete representation for bytes string
```

Comment:
This comment doesn't really tell us anything.

Reply:
Hmmm, it does though ^^" ? The `OuroborosApplication` has a parametric type here as `bytes`. That comment tells what the function of that `bytes` parameter. Two concrete instances are `ByteString` or `Text`, which then has an impact on the codec.